### PR TITLE
[TRTLLM-16283][feat] Prepare DeepSeek V4 attention for sparse KV offload

### PR DESCRIPTION
**Pull request by cascade812**
## Description

DeepSeek V4 ratio-4 sparse attention currently assumes compressed KV is GPU-resident. This PR adds the attention-side integration for keeping completed compressed-KV history on the host and fetching each layer's selected history pages back to the GPU. It prepares the KVCM fetch inputs and consumes the returned GPU page table through the existing sparse MLA/FMHA path.

The initial target is single-token ratio-4 decode with BF16 or per-tensor FP8 main KV. The feature is default-off, and enabling `enable_kv_cache_offload` still raises `NotImplementedError` until the KVCM runtime and end-to-end validation are complete.

## Workflow and PR scope

The diagram shows the full intended workflow: **blue = this PR**, **gray = existing execution**, and **orange = pending KVCM runtime work**.

image

| Stage in the diagram | Added by this PR (blue) |
|---|---|
| Declare ratio-4 compressed KV sparse | Add the opt-in configuration and separate sparse main `COMPRESS` buffers from resident indexer/SWA caches. |
| Prepare each batch | Wait for prior cache work, upload mixed-tier page tables in request order with `copy_base_page_indices_to_device()`, record history counts, and build the resident GPU write table in persistent metadata. |
| Decode: select and request history pages | After the indexer/compressor streams join, convert top-k token indices to unique history-page ordinals and call `fetch_sparse_pages()` for this layer's `COMPRESS` buffer. Repeat for each ratio-4 sparse layer. |
| Decode: consume fetched KV | Merge fetched history with the resident tail into a separate read table, validate required pages and bounds, mask padded rows, and convert the original top-k tokens to GPU indices for sparse MLA/FMHA. |
| After all model layers and sampling | Order cache work after model readers/writers, then use the existing `resize()` history update to make newly completed pages eligible for offload. The next batch waits for that cache work before refreshing its tables. |

Fresh prefill uses the resident write table and existing attention path without a history fetch. The existing compressor, indexer, and sparse MLA/FMHA kernels remain the compute path shown in gray.

The timing and data distinctions are:

- History becomes eligible for offload **after the whole model step**; selected history is refetched **per sparse layer** during decode.
- Page selection deduplicates transfers. Attention retains the original top-k token order, duplicates, and within-page positions.
- The compressor writes resident GPU pages through the write table; attention reads fetched history plus the resident tail through a separate read table.

**Pending KVCM work (orange):** allocate host storage, resident GPU pages, and fetch scratch; move newly completed history from GPU to host and safely release its GPU slots; copy selected host pages to GPU scratch and populate the fetched-page table. KVCM sparse API declarations and runtime implementation are separate work. Fetch-output units, scratch sizing/lifetime, and native offloaded-attention parity must be established before enabling inference. This PR also rejects unsupported batch layouts and history policies.

## Test Coverage

- The complete rerun passed 149 tests on H100 in 72.76 seconds, with no failures or skips. It covers every newly added or modified test function, plus configuration regressions and the golden manifest check.
- Coverage includes metadata preparation, page selection/merge, per-layer fetch/address consumption, history boundaries and stream ordering, required-page coverage, FP8 scheduler preparation, CUDA graph replay/padding, and rejection paths.
- Earlier B200 validation passed 120 offload tests in 34.85 seconds.
- Commit hooks, lint/formatting, Python compilation, and diff whitespace checks passed.

GPU validation used KVCM test doubles, real Torch/Triton kernels, and real Torch D2H/H2D copies in the stream-ordering test. The shared runtime/source overlay had native operator initialization disabled. Real KVCM transfers and native offloaded attention parity remain unvalidated.

## PR Checklist

- [x] New paths have focused tests covered by the existing CPU/B200 test lists.
- [x] Generated the LLM args golden manifest for the additive configuration field.
- [x] Telemetry/privacy CODEOWNER approval for the new public configuration field.
- [x] Confirm the implemented KVCM contracts and native attention correctness before enabling offload.
- [x] Please check this after reviewing the repository PR checklist as appropriate for this PR.

## Dev Engineer Review

- The supplied diff contains changes only in `tensorrt_llm/llmapi/llm_args.py` and `tests/unittest/llmapi/test_llm_args.py`.
- The configuration change adds or updates DeepSeek V4 sparse-attention argument handling. Verify default behavior, validation, parameter propagation, and JSON compatibility.
- No cache-manager, kernel, metadata, runtime, or KVCM implementation changes are present in the supplied diff.
- No review finding counts or test execution results were supplied.

## QA Engineer Review

- `tests/unittest/llmapi/test_llm_args.py` changes configuration tests for DeepSeek V4 sparse attention, including offload defaults, propagation, JSON round trips, ratio-4 validation, and positive `index_topk` validation.
- No integration test-list file changed. Existing DeepSeek V4 entries were found in `tests/integration/test_lists/test-db/l0_b200.yml`, but no new entry for these changes was supplied.
- Coverage verdict: **needs follow-up** until the updated unit tests run successfully and the relevant CI listing is confirmed.

## Per-File QA Perspective

- `tensorrt_llm/llmapi/llm_args.py`: Verify the `enable_kv_cache_offload` configuration default, validation conditions, and propagation to sparse and metadata parameters.
- `tests/unittest/llmapi/test_llm_args.py`: Covers configuration defaults, propagation, JSON serialization, and invalid ratio/top-k combinations. The test file is not shown as newly listed in `tests/integration/test_lists/test-db/` or `tests/integration/test_lists/qa/`.

### CODE REVIEW

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/backend.py [hunk 1]
```diff
@@ -280,6 +280,7 @@ def sparse_attn_predict(
             local_layer_idx, DeepseekV4AttentionType.SWA.value
         ]
 
+        active_request_count = None
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/backend.py:283]: What `active_request_count` does here?  Looks like it doesn't seem necessary. Can we use the `num_contexts` and `num_generations` in `metadata` to get this `active_request_count`?

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/backend.py [hunk 2]
```diff
@@ -292,6 +293,45 @@ def sparse_attn_predict(
                 topk_indices = sparse_backend_args.topk_indices
                 assert topk_indices is not None, "topk_indices is required when compress_ratio=4"
                 compressed_local_indices = topk_indices
+                if self.sparse_attention_config.enable_kv_cache_offload:
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/backend.py:296]: Most checks here are already guaranteed before the forward runs (metadata setup and the per-batch validation). Could we simplify this to `if self.sparse_attention_config.enable_kv_cache_offload and generation_only: block_table_compressed = fetch_sparse_read_table(...)`?

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/metadata.py [hunk 3]
```diff
@@ -513,9 +574,57 @@ def restore_after_draft_forward(self, saved_state: dict | None) -> None:
...
+        is supported. This runs during prepare(), outside capture/replay;
+        the per-layer path consumes the resulting phase and device metadata.
+        """
+        if not self.sparse_metadata_params.enable_kv_cache_offload:
+            return
+        state = self.sparse_offload_state
+        if state is not None:
+            state.prepared = False
+            state.is_prefill = False
+        if self.beam_width != 1 or self.max_draft_tokens or self.draft_kv_cache_manager is not None:
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/metadata.py:590]: There are some duplicated checks in `validate_sparse_offload_batch` and metadata setup. Suggest checking fixed configuration once in `__init__` (beam width, spec decoding/draft, block reuse, Helix, KV compression, dtype) and batch shape once in `validate_sparse_offload_batch` (no mixed batches, no chunked prefill).

File: tests/unittest/_torch/attention/sparse/deepseek_v4/test_deepseek_v4_offload_attention.py
Comment (lfr-0531) [inline tests/unittest/_torch/attention/sparse/deepseek_v4/test_deepseek_v4_offload_attention.py:1]: Could we keep a new `test_deepseek_v4_offload.py` for the offload feature, a new `test_deepseek_v4_kernels.py` for all of the dsv4 specific kernels, and move the remaining tests into the existing files (`test_deepseek_v4_cache_manager.py`)?

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/cache_manager.py [hunk 4]
```diff
@@ -1579,6 +1610,314 @@ def copy_batch_sliding_block_tables(
...
+            )
+        local_layers = [
+            layer
+            for layer in self.pp_layers
+            if self._compress_ratios[layer] == DEEPSEEK_V4_SPARSE_RATIO
+        ]
+        if not local_layers:
+            return {}
+        if not callable(getattr(self.impl, "is_sparse", None)) or not callable(
+            getattr(self.impl, "copy_base_page_indices_to_device", None)
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/cache_manager.py:1699]: It seems that there is no `is_sparse` and `copy_base_page_indices_to_device` in KVCM v2.

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/cache_manager.py [hunk 5]
```diff
@@ -1579,6 +1610,314 @@ def copy_batch_sliding_block_tables(
...
+                state.active_request_count,
+                compressed_lengths,
+                state.compress_read_table,
+                self.tokens_per_block // DEEPSEEK_V4_SPARSE_RATIO,
+                descriptor.page_index_upper_bound,
+                state.read_table_valid,
+            )
+            # A sentinel is address-safe but would silently drop required KV.
+            # Fail on device before conversion/FMHA, including during replay.
+            torch._assert_async(
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/cache_manager.py:1915]: `torch._assert_async` is a device-side assert, which kills the CUDA context and takes down the whole server over one bad request. Could this sit behind a debug flag, with a recoverable per-request error path by default?

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/kernels.py [hunk 6]
```diff
@@ -103,10 +109,11 @@ def _deepseek_v4_local_to_global_kernel(
     swa_block_ordinal = swa_local_idx // tokens_per_block_swa
     swa_token_in_block = swa_local_idx % tokens_per_block_swa
     swa_valid_block = swa_block_ordinal < max_blocks_swa
-    swa_full_mask = swa_valid_mask & swa_valid_block
+    swa_full_mask = live & swa_valid_mask & swa_valid_block
 
     swa_bt_ptr = block_table_swa_ptr + req * bt_swa_stride0 + swa_block_ordinal * bt_swa_stride1
     swa_page_index = tl.load(swa_bt_ptr, mask=swa_full_mask, other=0)
+    swa_full_mask = swa_full_mask & (swa_page_index >= 0)
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/kernels.py:116]: This `swa_page_index >= 0` mask also apply when offload is off, so this changes the default path. And also the following `compressed_full_mask`.

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/cache_manager.py [hunk 7]
```diff
@@ -1579,6 +1610,314 @@ def copy_batch_sliding_block_tables(
...
+                )
+            stream = torch.cuda.current_stream()
+            stream.wait_stream(self._stream)
+            # A queued H2D copy borrows the pinned host staging. Wait only for
+            # that upload before reusing its storage, never for the whole GPU.
+            if state.history_upload_pending:
+                state.history_upload_done.synchronize()
+            state.history_blocks_host.zero_()
+            for row, history in enumerate(histories):
+                state.history_blocks_host[row] = history
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/cache_manager.py:1813]: Filling `history_blocks_host` one element at a time in a Python loop adds host overhead on every step. Could we use a single vectorized copy instead, e.g. `host[:n].copy_(torch.tensor(histories, dtype=torch.int32))`?

**Review by Mgluhovskoi — APPROVED** [review]

**Review by mikeiovine — APPROVED** [review]
Looks like this one needs careful review from the attention and KV cache teams; stamp on behalf of the runtime team to reduce the number of approvals we need.

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/backend.py (hunk 1 above)
Comment (cascade812) [reply tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/backend.py]: This is the live-request count excluding CUDA graph padding. num_contexts + num_generations includes the padded rows, so it can differ—for example, three live requests in a four-row graph.
